### ISSUE TITLE
MemoryDiagnoser large object heap limitation should be documented

### ISSUE THREAD

**Issue by zivkan** [dotnet/BenchmarkDotNet#2474]
I was recently surprised by some benchmark results, and due to a lack of experience investigating Large Object Heap allocations, it took me considerable time to confirm that a benchmark where the `MemoryDiagnoser` was reporting much less memory than I was expecting.  I only discovered the LOH limitation in MemoryDiagnoser because I had enough experience to have a feeling that the results didn't seem quite right, and the time & motivation to investigate. I'm worried that other people might not notice.

I was also surprised that I couldn't find any other documentation or github issues on Large Object Heap, or LOH, allocations. I only found a single stack overflow question saying that .NET doesn't provide an API. It's fine that BDN can't report on what it can't measure, but when people like me are using it to measure minimal reproductions of code that is going to exist in a much larger system, then being aware of limitations in BDN's memory reporting is important to not be mislead. If memory allocations were the only focus of a benchmark, then maybe many developers would notice that the results are not quite right and would investigate further. But when there are many things going on, it's easy to miss, so a warning would really help bring the risk to mind.

Therefore, I think that:

1. There should be documentation on https://benchmarkdotnet.org/articles/configs/diagnosers.html, and maybe also on the API reference for `MemoryDiagnoserAttribute` (and maybe even Intellisense too, to maximise the chance people will see the warning).
2. When the benchmark run completes, and BDN outputs the results to the console, there is already a line `// * Diagnostic Output - MemoryDiagnoser *`.  I think it should include a line saying something like ".Large Object Heap allocations are not measured due to .NET API limitations". This would be enough to prompt me to think about whether my test might be reading a file that's 85k (or 42.5k) in size and therefore if I need to be more careful in interpreting the benchmark results.

**Comment by zivkan** [comment]
I haven't yet tested this with BDN, but for anyone looking for a way to detect LOH allocations, I couldn't find anything that works on Linux, but on Windows the .NET runtime reports Performance Counters, which also get reported as ETW events. I have a small proof of concept:

```csharp
internal class LohEventListener : EventListener
{
    double MaxValue = 0;

    protected override void OnEventSourceCreated(EventSource eventSource)
    {
        Console.WriteLine($"Event source created: {eventSource.Name}");
        if (eventSource.Name == "System.Runtime")
        {
            this.EnableEvents(eventSource, EventLevel.LogAlways);
        }
    }

    protected override void OnEventWritten(EventWrittenEventArgs eventData)
    {
        if (eventData.EventSource.Name == "System.Runtime"
            && eventData.EventName == "EventCounters")
        {
            var payload = (IDictionary<string, object>)eventData.Payload[0];
            var name = (string)payload["Name"];
            if (name == "loh-size")
            {
                var value = (double)payload["Mean"];
                if (value > MaxValue)
                {
                    MaxValue = value;
                }
            }
        }
    }
}
```

I'm hoping that if I instansiate the listener in the benchmark class' constructor, or in the `[GlobalSetup]` method, then I can check if the `MaxValue` is `> 0` in the `[GlobalCleanup]` method, and if so, throw an exception. Hopefully this allows my benchmark app to disallow any benchmark implementation that causes LOH allocations.

**Comment by zivkan** [comment]
Sorry, just looks like another failed analysis on my side. My entire perf investigation has been a disaster from the start 😦 

In my case, it's related to System.Text.Json, and when you have a custom `JsonConverter<T>` for the root object, STJ will load the entire json text into a single `byte[]` (starting from 8k, then resizing/doubling each time). This was completely unintuitive to me, but eventually I learned it causes large heap allocations for these `byte[]` arrays, although it makes sense, otherwise the converter needs to handle running out of data in the buffer before reaching the end of the object.

Digging into it now, STJ uses a shared array pool, and will therefore pool the array that is allocated on the LOH. Therefore BenchmarkDotNet is correctly detecting that each iteration of the benchmark is allocating a certain amount, because it's not doing a LOH allocation each iteration.  Unfortunately in my case doing the LOH in the first place is unacceptable, even if it's then available for other code via the shared pool.

Fortunately the event listener looking for LOH event counter events, that I described earlier, works on Windows, so in scenarios where LOH allocations are not acceptable, I now have a tool. It just took me a long time, and a lot of effort, to figure out.
